From 04055b9897820665aa48d0533da5d2631e348ef7 Mon Sep 17 00:00:00 2001
From: Haihua Hu <jared.hu@nxp.com>
Date: Wed, 21 Dec 2016 13:35:57 +0800
Subject: [PATCH] [MMFMWK-7426] Make qmlglsrc be compatible with GLES2.0

Fallback to glCopyTexImage2D if glBlitFrameBuffer is not available
to avoid segfault

Upstream Status: Backport[1.11.1]

https://bugzilla.gnome.org/show_bug.cgi?id=777078

Signed-off-by: Haihua Hu <jared.hu@nxp.com>
---
 ext/qt/qtwindow.cc | 32 +++++++++++++++++---------------
 1 file changed, 17 insertions(+), 15 deletions(-)

diff --git a/ext/qt/qtwindow.cc b/ext/qt/qtwindow.cc
index 498416a..8a9af68 100644
--- a/ext/qt/qtwindow.cc
+++ b/ext/qt/qtwindow.cc
@@ -223,23 +223,25 @@ QtGLWindow::afterRendering()
   GST_DEBUG ("qml render target id %d, render to tex %d %dX%d", 
       this->source->renderTargetId(), dst_tex, width,height);
 
-  gl->BindFramebuffer (GL_DRAW_FRAMEBUFFER, this->priv->fbo);
   gl->BindTexture (GL_TEXTURE_2D, dst_tex);
-  gl->FramebufferTexture2D (GL_DRAW_FRAMEBUFFER, GL_COLOR_ATTACHMENT0,
-      GL_TEXTURE_2D, dst_tex, 0);
-
-  ret = gst_gl_context_check_framebuffer_status (context);
-  if (!ret) {
-    GST_ERROR ("FBO errors");
-    goto errors;
+  if (gl->BlitFramebuffer) {
+    gl->BindFramebuffer (GL_DRAW_FRAMEBUFFER, this->priv->fbo);
+    gl->FramebufferTexture2D (GL_DRAW_FRAMEBUFFER, GL_COLOR_ATTACHMENT0,
+        GL_TEXTURE_2D, dst_tex, 0);
+
+    ret = gst_gl_context_check_framebuffer_status (context);
+    if (!ret) {
+      GST_ERROR ("FBO errors");
+      goto errors;
+    }
+
+    gl->ReadBuffer ( GL_COLOR_ATTACHMENT0 );
+    gl->BlitFramebuffer ( 0, 0, width, height,
+                          0, 0, width, height,
+                          GL_COLOR_BUFFER_BIT, GL_LINEAR );
+  } else {
+    gl->CopyTexImage2D (GL_TEXTURE_2D, 0, GL_RGBA, 0, 0, width, height, 0);
   }
-
-  gl->ReadBuffer ( GL_COLOR_ATTACHMENT0 );
-  gl->BlitFramebuffer ( 0, 0, width, height,
-		        0, 0, width, height,
-		        GL_COLOR_BUFFER_BIT, GL_LINEAR );
-
-  //gl->CopyTexImage2D (GL_TEXTURE_2D, 0, GL_RGBA, 0, 0, width, height, 0);
   
   GST_DEBUG ("rendering finished");
 
-- 
1.9.1

